[Mobile] - Draggable component

geriux opened the pull request:
- `Gutenberg Mobile PR` -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/4675

Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4619

## What?
This PR adds the `Draggable` component which allows enabling draggable touch events to an element.

## Why?
This component is needed for the Drag & Drop blocks project.

## How?
By adding a new component that uses the libraries Reanimated and Gesture Handler to be able to detect long-press and panning events on an element.

## Testing Instructions

<details><summary>Apply this patch</summary>
<p>

```patch
diff --git a/packages/block-editor/src/components/block-list/block-list-item-cell.native.js b/packages/block-editor/src/components/block-list/block-list-item-cell.native.js
index c399643a63..9245077313 100644
--- a/packages/block-editor/src/components/block-list/block-list-item-cell.native.js
+++ b/packages/block-editor/src/components/block-list/block-list-item-cell.native.js
@@ -2,11 +2,17 @@
  * External dependencies
  */
 import { View } from 'react-native';
+import {
+	useSharedValue,
+	useAnimatedStyle,
+	withSpring,
+} from 'react-native-reanimated';
 
 /**
  * WordPress dependencies
  */
 import { useEffect, useCallback } from '@wordpress/element';
+import { Draggable } from '@wordpress/components';
 
 /**
  * Internal dependencies
@@ -15,6 +21,32 @@ import { useBlockListContext } from './block-list-context';
 
 function BlockListItemCell( { children, clientId, rootClientId } ) {
 	const { blocksLayouts, updateBlocksLayouts } = useBlockListContext();
+	const isDragging = useSharedValue( false );
+
+	const wrapperAnimatedStyles = useAnimatedStyle( () => {
+		return {
+			transform: [ { scale: withSpring( isDragging.value ? 0.95 : 1 ) } ],
+			backgroundColor: isDragging.value ? 'gray' : 'transparent',
+			opacity: withSpring( isDragging.value ? 0.5 : 1 ),
+		};
+	} );
+
+	const onDragStart = useCallback( ( ev ) => {
+		'worklet';
+		isDragging.value = true;
+		console.log( 'onDragStart', ev );
+	}, [] );
+
+	const onDragOver = useCallback( ( ev ) => {
+		'worklet';
+		console.log( 'onDragOver', ev );
+	}, [] );
+
+	const onDragEnd = useCallback( () => {
+		'worklet';
+		isDragging.value = false;
+		console.log( 'onDragEnd' );
+	}, [] );
 
 	useEffect( () => {
 		return () => {
@@ -36,7 +68,16 @@ function BlockListItemCell( { children, clientId, rootClientId } ) {
 		[ clientId, rootClientId, updateBlocksLayouts ]
 	);
 
-	return <View onLayout={ onLayout }>{ children }</View>;
+	return (
+		<Draggable
+			wrapperAnimatedStyles={ wrapperAnimatedStyles }
+			onDragStart={ onDragStart }
+			onDragEnd={ onDragEnd }
+			onDragOver={ onDragOver }
+		>
+			<View onLayout={ onLayout }>{ children }</View>
+		</Draggable>
+	);
 }
 
 export default BlockListItemCell;

```

</p>
</details>

- Open the demo app
- Add a few blocks
- Open the console log
- Long press on a block
- Expect to see the logs on the following events: `onDragStart`, `onDragEnd`, and `onDragOver` with the coordinates data.

## Screenshots or screencast <!-- if applicable -->

<kbd><img src="https://user-images.githubusercontent.com/4885740/159249277-54792fb5-7eb1-4672-ab17-3249720ff476.gif" /></kbd>

Commit:
Mobile - Add Draggable component

Commit:
Merge branch 'rnmobile/feature/drag-and-drop' into rnmobile/feature/drag-and-drop-draggable-component

Commit:
Mobile - Draggable - Fix composition of gestures, enable Pan gesture once LongPress is recognized

Commit:
Mobile - Draggable component - Track if the gesture Pan started, if it didn't and long pressure was activated it should call onDragEnd

Commit:
Mobile - Draggable component - Add props documentation

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+ * @param {?number}                                     props.minDuration           Minimum time, that a finger must remain pressed on the corresponding view.
+ * @param {Function}                                    props.onDragEnd             Callback when dragging happens over an element.
+ * @param {Function}                                    props.onDragOver            Callback when dragging ends.
+ * @param {Function}                                    props.onDragStart           Callback when dragging stars.
```
Typo:
```suggestion
 * @param {Function}                                    props.onDragStart           Callback when dragging starts.
```

geriux replied:
😅 updated in https://github.com/WordPress/gutenberg/pull/39551/commits/d8023e6d675ed58843275f865fb580fc402c4f9b#diff-4c96695e2f9beedb8860ca314946837c593df76d021a9167b7f90738ddcaabf0R16

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+ * @param {?number}                                     props.maxDistance           Maximum distance, that defines how far the finger is allowed to travel during a long press gesture.
+ * @param {?number}                                     props.minDuration           Minimum time, that a finger must remain pressed on the corresponding view.
+ * @param {Function}                                    props.onDragEnd             Callback when dragging happens over an element.
+ * @param {Function}                                    props.onDragOver            Callback when dragging ends.
```
I think the definitions of these two callbacks are mismatching, as `onDragEnd` should be called when dragging ends.
```suggestion
 * @param {Function}                                    props.onDragEnd             Callback when dragging ends.
 * @param {Function}                                    props.onDragOver            Callback when dragging happens over an element.
```

geriux replied:
Thanks for spotting this, updated in https://github.com/WordPress/gutenberg/pull/39551/commits/d8023e6d675ed58843275f865fb580fc402c4f9b#diff-4c96695e2f9beedb8860ca314946837c593df76d021a9167b7f90738ddcaabf0R14-R15

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+ * @param {Object}                                      props                       Component props.
+ * @param {JSX.Element}                                 props.children              Children to be rendered.
+ * @param {?number}                                     props.maxDistance           Maximum distance, that defines how far the finger is allowed to travel during a long press gesture.
+ * @param {?number}                                     props.minDuration           Minimum time, that a finger must remain pressed on the corresponding view.
```
If these props are optional we should reflect them in the documentation, i.e.:
```suggestion
 * @param {?number}                                     [props.maxDistance]           Maximum distance, that defines how far the finger is allowed to travel during a long press gesture.
 * @param {?number}                                     [props.minDuration]           Minimum time, that a finger must remain pressed on the corresponding view.
```

**NOTE:** We could apply this to the callbacks as they're optional too.

geriux replied:
Updated in https://github.com/WordPress/gutenberg/pull/39551/commits/d8023e6d675ed58843275f865fb580fc402c4f9b#diff-4c96695e2f9beedb8860ca314946837c593df76d021a9167b7f90738ddcaabf0R12-R13

geriux replied:
>NOTE: We could apply this to the callbacks as they're optional too.

Good idea, I added them here https://github.com/WordPress/gutenberg/pull/39551/commits/d8023e6d675ed58843275f865fb580fc402c4f9b#diff-4c96695e2f9beedb8860ca314946837c593df76d021a9167b7f90738ddcaabf0R14-R16

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+ * @param {Function}                                    props.onDragEnd             Callback when dragging happens over an element.
+ * @param {Function}                                    props.onDragOver            Callback when dragging ends.
+ * @param {Function}                                    props.onDragStart           Callback when dragging stars.
+ * @param {import('react-native-reanimated').StyleProp} props.wrapperAnimatedStyles Animated styles for the wrapper component.
```
If this prop is optional, we could note it in the documentation:
```suggestion
 * @param {import('react-native-reanimated').StyleProp} [props.wrapperAnimatedStyles] Animated styles for the wrapper component.
```

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+		.shouldCancelWhenOutside( false );
+
+	const panGesture = Gesture.Pan()
+		.manualActivation( true )
+		.onTouchesMove( ( _, state ) => {
+			'worklet';
+			if ( isDragging.value ) {
+				hasPanStarted.value = true;
+				state.activate();
+			} else {
+				state.fail();
+			}
+		} )
```
I'm not sure if we need this if we already defined that this gesture event should work simultaneously with the long press (i.e. `simultaneousWithExternalGesture( longPressGesture )`). I understand that setting that relation means that the pan gesture will only turn active if the long press gesture is also active, right?

fluiddot replied:
When testing on Android the component `BlockDraggable` I'm working on, I found that this part is required to allow scrolling the block list when tapping over the blocks. Therefore, these findings might invalidate my previous comment.

geriux replied:
With the [latest comment](https://github.com/WordPress/gutenberg/pull/39551#discussion_r831390288) since it changes the composition to `Simultaneous` and `shouldCancelWhenOutside` I wonder if it still requires `simultaneousWithExternalGesture( longPressGesture )` I deleted it in the last change but I can bring it back if it affects scrolling the block list.

fluiddot replied:
Yeah, makes sense that `simultaneousWithExternalGesture ` statement is no longer quired due to using `Simultaneous` composition 👍 .

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+			}
+		} )
+		.onEnd( () => {
+			if ( ! hasPanStarted.value && isDragging.value ) {
```
Probably we don't need to check `isDragging.value` here as [`onEnd`](https://docs.swmansion.com/react-native-gesture-handler/docs/api/gestures/long-press-gesture#onendcallback) should only be called when the long press gesture is active, this means if `onStart` was already called (when it turned active).

geriux replied:
If we remove it, `onDragEnd` gets called two times if both Gestures are active, since I started with only `LongPress` and then introduced `Pan` and then it changed to manually activating the `Pan` gesture, I think we could remove the `onEnd` from the `Pan` gesture and only have it for the `LongPress` gesture. With this we could also remove the `hasPanStarted` shared value to simplify things. What do you think?

fluiddot replied:
Good point, I've just checked that both `onEnd` callbacks are called when releasing the touch. In that case, I agree with what you suggested about calling `onDragEnd` only in the `LongPress` gesture, and removing `onEnd` in the `Pan` gesture 👍 .

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+		} )
+		.onUpdate( ( ev ) => {
+			'worklet';
+			if ( isDragging.value ) {
```
I think we don't need to check this condiction because [`onUpdate`](https://docs.swmansion.com/react-native-gesture-handler/docs/api/gestures/pan-gesture#onupdatecallback) is only called if long-press and pan gestures are both active.

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+		.onEnd( () => {
+			'worklet';
+
+			if ( isDragging.value && onDragEnd ) {
```
Similar to previous comments, probably it's not necessary to check `isDragging.value` as [`onEnd`](https://docs.swmansion.com/react-native-gesture-handler/docs/api/gestures/pan-gesture#onendcallback) callback is only called when long-press and pan gestures are active.

fluiddot reviewed (commented):
Awesome work @geriux 🏅 ! I tested the changes with both Android and iOS demo apps and worked as expected 🎊.

Before approval, it would be great if you could check some of the comments I added related to the documentation and potential improvements in the gesture logic, thanks 🙇 !

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+			isDragging.value = false;
+			hasPanStarted.value = false;
+		} )
+		.simultaneousWithExternalGesture( longPressGesture );
```
While testing these changes with the component `BlockDraggable` that I'm working on, I noticed that on Android `onDragOver` callback is not being called. I managed to solve it by adding to the pan gesture `.shouldCancelWhenOutside( false )`.

geriux replied:
Interesting! Yeah according to the docs:

> When true the handler will cancel or fail recognition (depending on its current state) whenever the finger leaves the area of the connected view

For the `Pan` gesture is `true` by default, and since we are leaving the connected view _behind_ setting it to `false` is required.

I'll add it as well, thanks for letting me know!

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,94 @@
+				hasPanStarted.value = true;
+				state.activate();
+			} else {
+				state.fail();
```
After further testing these changes and `BlockDraggable` component, I noticed that on Android the callback `onTouchesMove` can be called before `LongPress.onStart`. Besides, setting the state of the pan gesture to `failed` prevents this callback to be triggered again until releasing the touch. All these lead to a situation where the draggable gesture doesn't work if the previous is fulfilled, which as far as I tested can be easily reproduced.

Being that said, I think we might consider dropping setting the state to `failed` to prevent the issue described above.

fluiddot replied:
Btw, I managed to fix this by applying the following patch:
```diff
diff --git a/packages/components/src/draggable/index.native.js b/packages/components/src/draggable/index.native.js
index a9543a0d8fd87ea38c4ff07fce66b67bf103ad2b..85037fe80d71e4503feb737195ab2b70f753d10c 100644
--- a/packages/components/src/draggable/index.native.js
+++ b/packages/components/src/draggable/index.native.js
@@ -59,8 +59,6 @@ export default function Draggable( {
 			if ( isDragging.value ) {
 				hasPanStarted.value = true;
 				state.activate();
-			} else {
-				state.fail();
 			}
 		} )
 		.onUpdate( ( ev ) => {
@@ -80,9 +78,10 @@ export default function Draggable( {
 			isDragging.value = false;
 			hasPanStarted.value = false;
 		} )
+		.shouldCancelWhenOutside( false )
 		.simultaneousWithExternalGesture( longPressGesture );
 
-	const dragHandler = Gesture.Race( panGesture, longPressGesture );
+	const dragHandler = Gesture.Simultaneous( panGesture, longPressGesture );
 
 	return (
 		<GestureDetector gesture={ dragHandler }>

```

geriux replied:
Nice findings! I will apply the mentioned changes 👍

geriux replied:
I'm seeing that removing `state.fail();` causes the scroll on iOS to not work 😅 . Can you reproduce this? On Android the scroll keeps working.

Maybe we could only add `state.fail()` for iOS 🤔

fluiddot replied:
> I'm seeing that removing `state.fail();` causes the scroll on iOS to not work 😅 . Can you reproduce this? On Android the scroll keeps working.
> 
> Maybe we could only add `state.fail()` for iOS 🤔

Right, I thought I tested on iOS but I've just reproduced the same you mentioned. At this point, I think the better would be to check the platform and call `state.fail()` only for the iOS case.

geriux replied:
Cool, I'll add the iOS check 👍

Commit:
Mobile - Draggable component - Update documentation

Commit:
Mobile - Draggable component - Refactor that includes:
- Usage of only one onEnd callback for both gestures.
- Removes the hasPanStarted value since the onEnd callback is unified
- Adds shouldCancelWhenOutside for the Pan gesture.
- Removes the simultaneousWithExternalGesture since the Pan gesture is manually activated and composed with the LongPress gesture.
- Add isIOS check within onTouchesMove for the state.fail() call.

geriux commented:
Hey @fluiddot 👋 I've updated the code with the latest feedback, can you check it out when you can? Thanks!

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -0,0 +1,80 @@
+/**
+ * External dependencies
+ */
+import { Platform } from 'react-native';
```
We might consider using the `Platform` object from `@wordpress/element` package ([reference](https://github.com/WordPress/gutenberg/blob/trunk/packages/element/src/platform.ios.js)).

geriux replied:
Thanks for the suggestion! Updated in https://github.com/WordPress/gutenberg/pull/39551/commits/9afd4b911dc46fa3bd0422c25cf2d5ac6d38dcfd

fluiddot reviewed (approved):
LGTM 🎊 !

> Hey @fluiddot 👋 I've updated the code with the latest feedback, can you check it out when you can? Thanks!

I tested these changes with the WIP component `BlockDraggable` and work perfectly.

Commit:
Mobile - Draggable component - Use Platform from @wordpress/element
